gadget/update.go

```diff
@@ -20,8 +20,102 @@ package gadget
 
 import (
 	"errors"
+	"fmt"
 )
 
 var (
 	ErrNoUpdate = errors.New("no update needed")
 )
+
+func resolveVolume(old *Info, new *Info) (oldVol, newVol *Volume, err error) {
+	// TODO: support only one volume
+	if len(new.Volumes) != 1 || len(old.Volumes) != 1 {
+		return nil, nil, errors.New("cannot update with more than one volume")
+	}
+
+	var name string
+	for n := range old.Volumes {
+		name = n
+		break
+	}
+	oldV := old.Volumes[name]
+
+	newV, ok := new.Volumes[name]
+	if !ok {
+		return nil, nil, fmt.Errorf("cannot find entry for volume %q in updated gadget info", name)
+	}
+
+	return &oldV, &newV, nil
+}
+
+func isSameOffset(one *Size, two *Size) bool {
+	if one == nil && two == nil {
+		return true
+	}
+	if one != nil && two != nil {
+		return *one == *two
+	}
+	return false
+}
+
+func isSameRelativeOffset(one *RelativeOffset, two *RelativeOffset) bool {
+	if one == nil && two == nil {
+		return true
+	}
+	if one != nil && two != nil {
+		return *one == *two
+	}
+	return false
+}
+
+func isLegacyMBRTransition(from *PositionedStructure, to *PositionedStructure) bool {
+	// legacy MBR could have been specified by setting type: mbr, with no
+	// role
+	return from.Type == MBR && to.EffectiveRole() == MBR
```

Comment:
What if from.Type is MBR but from.Role is set to something else? Or this will never happen?

Reply:
Hm this should be caught by validation. Let me double check if it is. If not, I'll open a separate PR with the fix.

Reply:
Added in a separate PR: #6832